modules/webgl/src/context/context.js

```diff
@@ -203,3 +196,30 @@ function getVersion(gl) {
   // Must be a WebGL1 context.
   return 1;
 }
+
+// use devicePixelRatio to set canvas width and height
+function setDevicePixelRatio(gl, devicePixelRatio) {
+  let devicePixelRatioClamped = false;
+  let aspectRatioValid = false;
+  let clientSizeValid = false;
+
+  // Note: when devicePixelRatio is too high, it is possible we might hit system limit for
+  // drawing buffer width and hight, in those cases they get clamped and resulting aspect ration may not be maintained
+  // for those cases, reduce devicePixelRatio.
+  do {
+    const {clientWidth, clientHeight} = gl.canvas;
+    gl.canvas.width = Math.ceil(clientWidth * devicePixelRatio);
+    gl.canvas.height = Math.ceil(clientHeight * devicePixelRatio);
```

Comment:
This no longer works if `clientWidth` is not valid?

Reply:
I am using `clientSizeValid` variable , if `clientWidth` or `clientHeight` is not valid, we exit the loop and the behavior remains same.

Reply:
Event if `clientSizeValid` is `false` you have already overwritten the canvas width and height.

Reply:
we were doing the same before these changes, I can set a safe value , may be `1` if width or height not valid ?

Reply:
Added safe back up value of 1.